duckduckgo/Android
# Add app friction survey

Task/Issue URL: https://app.asana.com/0/361428290920652/885889523162856
Tech Design URL: https://app.asana.com/0/361428290920652/892077194142864

**Description**:
Adds an app friction survey to the app, driven by survey-side configuration.

**Steps to test this PR**:

**TEST ONE: Survey appears until completed**
- Install a fresh version of the app
- Using a tool such as Charles intercept the survey request to https://staticcdn.duckduckgo.com/survey/survey-mobile.json and replace with
```
{
  "id": "509c0d52-f4a8-4409-a9b1-d858af42fa01",
  "surveyOptions": [
    {
      "url": "https://www.surveymonkey.com/r/9JTMFGJ",
      "installationDay": 0,
      "ratioOfUsersToShow": 1
    }
  ]
}
```
- Ensure that the survey CTA box appears
- Quit the app and reopen, the survey should still be there
- Open, fill in the survey and submit it
- Ensure the survey disappears
- Kill the app and sync again, providing the same json as before
- Ensure the survey does not again appear

**TEST TWO: Survey Dismissed**
- Again, intercept the survey request and replace with a new survey
```
{
  "id": "509c0d52-f4a8-4409-a9b1-d858af42fa02",
  "surveyOptions": [
    {
      "url": "https://www.surveymonkey.com/r/9JTMFGJ",
      "installationDay": 0,
      "ratioOfUsersToShow": 1
    }
  ]
}
```
- Dismiss the survey by pressing the "x" button in the CTA
- Ensure the survey disappears
- Quit the app and reopen, the survey CTA should not appear again

**TEST THREE: Survey fails to load**
- Again, intercept the survey request and replace with a new survey
```
{
  "id": "509c0d52-f4a8-4409-a9b1-d858af42fa03",
  "surveyOptions": [
    {
      "url": "https://www.surveymonkey.com/r/9JTMFGJ",
      "installationDay": 0,
      "ratioOfUsersToShow": 1
    }
  ]
}
```
- Switch on airplane mode on the device
- Open the survey
- Note the error message appears
- Close the survey and note that the CTA is still there

**TEST FOUR: Switch survey off**
- As before, intercept the survey request and respond with a new survey id
```
{
  "id": "509c0d52-f4a8-4409-a9b1-d858af42fa04",
  "surveyOptions": [
    {
      "url": "https://www.surveymonkey.com/r/9JTMFGJ",
      "installationDay": 0,
      "ratioOfUsersToShow": 1
    }
  ]
}
```
- Note that the survey is displayed
- Now run the app again this time responding with an empty response
```
{}
```
- Note that the survey is no longer displayed

**TEST FIVE: Allocation**
- Run a fresh install of the app, multiple times, without intercepting requests
- Note that the survey is not shown
- Fast-forward the devices date by 1, 7 and 10 days
- Note that the app is sometimes shown on these days depending on survey allocation

**TEST SIX: Database Migration**
- Install previous app version and run the app
- Install this version over the top and ensure nothing goes "boom"!

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
